### DISCUSSION TITLE
Working home row mods! 👍🏻  (And some tidbits.)

### DISCUSSION THREAD

**Discussion by Erlendms** [evan-liu/karabiner.ts#118] [Show and tell]
Hi!
I've been working on seeing if I can make good [Home row mods](https://precondition.github.io/home-row-mods) with Karabiner (as opposed to QMK). **And with the help of karabiner.ts, I think I've succeeded!**

Info, and .jsons to copy, can be found [in this repo](https://github.com/Erlendms/karabiner-actions.git).

I used the method that's been highlighted elsewhere, so the individual mods are made like this:

```      
map("a")
  .toIfAlone("a", {}, { halt: true })
  .toIfHeldDown("l⇧", {}, { halt: true })
  .toDelayedAction(toKey("vk_none"), toKey("a")),
```

I also found a way to mimic something close to Permissive hold from QMK. But I don't love that feature in the first place. Secondly, if you trigger the mod with this method, you can't use the same mod for more clicks afterwards.

But perhaps someone else can look at what I got so far and improve it!

```
map("a")
  .toIfAlone("a", {}, { halt: true })
  .toDelayedAction(
   toKey("vk_none"),
  toStickyModifier("left_shift", "toggle"),
 )
.toIfHeldDown("l⇧", {}, { halt: true }),
```
(So, I changed what happens if you interrupt, to be a sticky shift.)

---

I also made a small improvement (IMO) to the classic _Caps Lock ➔ Hyper key_ formula. 
Because, I like to have the Caps Lock function available - but too often I will press it to use it for Hyper, but then change my mind and not hit a hotkey. Normally this would trigger Caps Lock (since it was alone) - **but with my new version, it's only Caps Lock if it's alone _and_ you hold it under the `to_if_held_down_treshold`.**

I couldn't find a way to do it 100 % in _karabiner.ts_. There I did this (the parameters isn't needed if you don't want it to be longer than your regular one):
```
map("⇪")
  .toIfAlone("⇪", {}, { halt: true })
  .toDelayedAction(
  toKey("vk_none"),
  toStickyModifier("left_shift", "toggle"),
  )
  .toIfHeldDown("l⇧", "l⌘⌥⌃", { halt: true })
  .parameters({
  "basic.to_if_held_down_threshold_milliseconds": 300,
  "basic.to_delayed_action_delay_milliseconds": 350,
}),
```

But then I had to add the other modifiers in the .json afterwards. **Is there a way to add them in karabiner.ts?**

Here you can see I've added the other modifiers under "left_shift". 👇🏻 
```
"to_delayed_action": {
        "to_if_canceled": [
          {
            "sticky_modifier": {
              "left_shift": "toggle"
            }
          },
          {
            "sticky_modifier": {
              "left_control": "toggle"
            }
          },
          {
            "sticky_modifier": {
              "left_option": "toggle"
            }
          },
          {
            "sticky_modifier": {
              "left_command": "toggle"
            }
          }
        ],
        "to_if_invoked": [
          {
            "key_code": "vk_none"
          }
        ]
      },
```

---

Anyway! I hope someone stumbles over this while searching for "Home row mods Karabiner", like I've done for a couple of weeks now, and get some enjoyment out of this. 🙂 

-Erlend

**Comment by evan-liu** [comment]
Hi Erlend, that's very cool, thanks for sharing!

Any valid karabiner config JSON should be able to write in `karabiner.ts`, could you please try if this works for you: 

```typescript
map('⇪')
  .toIfAlone('⇪')
  .toDelayedAction(
    toNone(),
    [
      toStickyModifier('left_shift', 'toggle'),
      toStickyModifier('left_control', 'toggle'),
      toStickyModifier('left_option', 'toggle'),
      toStickyModifier('left_command', 'toggle'),
    ]
  )
  .toIfHeldDown('l⇧', 'l⌘⌥⌃'),
``` 

Or a bit shorter: 

```typescript
map('⇪')
  .toIfAlone('⇪')
  .toDelayedAction(
    toNone(),
    ['shift', 'control', 'option', 'command'].map((mod) =>
      toStickyModifier(`left_${mod}`, 'toggle')
    )
  )
  .toIfHeldDown('l⇧', 'l⌘⌥⌃'),
```

**Reply by Erlendms** [reply]
Ah, that worked! Thanks! 🙌🏻 
What I missed, was the square brackets around the multiple stickyModifiers... (Which I should've remembered, as I noticed that was a pattern! 🤦🏻‍♂️)

---

A couple of issues I've had, that I don't know if there's any answers two:

1. Using the toDelayedAction (with stickyModifier) to simulate _Permissive Hold_ made me unable to keep holding the modifier.

2. I found myself wanting the delayedLayer function [discussed here](https://github.com/evan-liu/karabiner.ts/discussions/114)... 😊 I wanted to make a navigation layer accessed via letters - but as I couldn't use the toDelayedAction method of getting rolls to type out the letters, I kept missing key pressed. 😕

**Reply by evan-liu** [reply]
I would also love to use `delayedLayer` if there is a way to make it work. Maybe someday, Karabiner could add more features or enhance `toDelayedAction`.

For now, I am using [Duo Modifiers](https://karabiner.ts.keymaps.app/examples/modifier-keys/duo-modifiers) — more keys to press but fewer accidents.
